File: src/ingest.ts
import { LogPayload, WatchOptions } from './types';

const DEFAULT_ENDPOINT = 'https://api.llmwatch.io/ingest';

// Printed once per process so dev mode is obvious but not spammy.
let devNoticePrinted = false;

function buildEnriched(payload: LogPayload, options: WatchOptions): Record<string, unknown> {
  const out: Record<string, unknown> = { ...payload };
  if (options.projectId)    out.projectId = options.projectId;
  if (options.userId)       out.userId    = options.userId;
  if (options.sessionId)    out.sessionId = options.sessionId;
  if (options.tags?.length) out.tags      = options.tags;
  if (options.metadata)     Object.assign(out, options.metadata);
  return out;
}

function truncate(s: string, n = 120): string {
  return s.length > n ? `${s.slice(0, n)} …` : s;
}

function devLog(payload: LogPayload, options: WatchOptions): void {
  if (!devNoticePrinted) {
    console.log('[llmwatch] Dev mode — set apiKey to send logs to the dashboard.\n');
    devNoticePrinted = true;
  }

  const enriched = buildEnriched(payload, options);
  const cost = payload.cost_usd > 0 ? `$${payload.cost_usd.toFixed(6)}` : 'cost unknown';

  const meta = [
    payload.model,
    `${payload.latency_ms}ms`,
    `${payload.tokens_used} tokens`,
    cost,
    enriched.projectId && `project:${enriched.projectId}`,
    enriched.userId    && `user:${enriched.userId}`,
    enriched.sessionId && `session:${enriched.sessionId}`,
  ].filter(Boolean).join('  ·  ');

  console.log(`[llmwatch] ${meta}`);
  console.log(`  prompt:     ${truncate(payload.prompt)}`);
  console.log(`  completion: ${truncate(payload.completion)}`);
  if (Array.isArray(enriched.tags) && enriched.tags.length) {
    console.log(`  tags:       ${enriched.tags.join(', ')}`);
  }
  console.log('');
}

export function sendLog(payload: LogPayload, options: WatchOptions): void {
  if (!options.apiKey) {
    devLog(payload, options);
    return;
  }

  const body = JSON.stringify(buildEnriched(payload, options));

  // Intentionally fire-and-forget: ingest must never throw or block the caller.
  fetch(options.endpoint ?? DEFAULT_ENDPOINT, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${options.apiKey}`,
      'User-Agent': 'llmwatch-sdk/0.4.0',
    },
    body,
  }).then(res => {
    if (!res.ok) {
      options.onError?.(
        new Error(`LLMWatch ingest failed: HTTP ${res.status} ${res.statusText}`),
      );
    }
  }).catch(err => {
    options.onError?.(err instanceof Error ? err : new Error(String(err)));
  });
}

File: python-sdk/loglens/_ingest.py
import json
import threading
import urllib.request
import urllib.error
from dataclasses import asdict

DEFAULT_ENDPOINT = 'https://llmwatch-production.up.railway.app/ingest'
_dev_notice_printed = False


def _do_send(body: bytes, endpoint: str, api_key: str, on_error) -> None:
    try:
        req = urllib.request.Request(
            endpoint,
            data=body,
            headers={
                'Content-Type': 'application/json',
                'Authorization': f'Bearer {api_key}',
                'User-Agent': 'loglens-sdk-python/0.1.1',
            },
            method='POST',
        )
        with urllib.request.urlopen(req, timeout=10) as resp:
            if resp.status >= 400 and on_error:
                on_error(Exception(f'LogLens ingest failed: HTTP {resp.status}'))
    except Exception as e:
        if on_error:
            on_error(e)


def _truncate(s: str, n: int = 120) -> str:
    return s[:n] + ' …' if len(s) > n else s


def _dev_log(payload, options) -> None:
    global _dev_notice_printed
    if not _dev_notice_printed:
        print('[loglens] Dev mode — set api_key to send logs to the dashboard.\n')
        _dev_notice_printed = True

    if payload.error_type:
        print(f'[loglens] {payload.model}  ·  {payload.latency_ms}ms  ·  ERROR: {payload.error_type}')
        print(f'  prompt:  {_truncate(payload.prompt)}')
        print(f'  error:   {payload.error_message or ""}')
        print()
        return

    cost_str = f'${payload.cost_usd:.6f}' if payload.cost_usd > 0 else 'cost unknown'
    parts = [payload.model, f'{payload.latency_ms}ms', f'{payload.tokens_used} tokens', cost_str]
    if options.project_id:
        parts.append(f'project:{options.project_id}')
    print(f'[loglens] {"  ·  ".join(parts)}')
    print(f'  prompt:     {_truncate(payload.prompt)}')
    print(f'  completion: {_truncate(payload.completion)}')
    if options.tags:
        print(f'  tags:       {", ".join(options.tags)}')
    print()


def send_log(payload, options) -> None:
    if not options.api_key:
        _dev_log(payload, options)
        return

    endpoint = options.endpoint or DEFAULT_ENDPOINT
    body_dict = asdict(payload)

    # remove None values and add enrichment fields
    body_dict = {k: v for k, v in body_dict.items() if v is not None}
    if options.mask_prompts:
        body_dict['prompt']     = '[masked]'
        body_dict['completion'] = '[masked]'
    if options.project_id:
        body_dict['projectId'] = options.project_id
    if options.user_id:
        body_dict['userId'] = options.user_id
    if options.session_id:
        body_dict['sessionId'] = options.session_id
    if options.tags:
        body_dict['tags'] = options.tags

    body = json.dumps(body_dict).encode('utf-8')

    t = threading.Thread(
        target=_do_send,
        args=(body, endpoint, options.api_key, options.on_error),
        daemon=True,
    )
    t.start()
